// qubic-ts-library/dist/qubic-types/transacion-payloads/QubicTransferAssetPayload.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.QubicTransferAssetPayload = void 0;
const QubicPackageBuilder_1 = require("../../QubicPackageBuilder");
const DynamicPayload_1 = require("../DynamicPayload");
const Long_1 = require("../Long");
const PublicKey_1 = require("../PublicKey");
/**
 *
 * Transaction Payload to transfer an Asset
 *
 * typedef struct
* {
*     uint8_t issuer[32];
*     uint8_t newOwnerAndPossessor[32];
*     unsigned long long assetName;
*     long long numberOfUnits;
* } TransferAssetOwnershipAndPossession_input;
 *
 *
 *
 */
class QubicTransferAssetPayload {
    constructor() {
        this._internalPackageSize = 32 + 32 + 8 + 8;
    }
    // todo: think about adding getters
    setIssuer(issuer) {
        if (typeof issuer === "string") {
            this.issuer = new PublicKey_1.PublicKey(issuer);
        }
        else {
            this.issuer = issuer;
        }
        return this;
    }
    setNewOwnerAndPossessor(newOwnerAndPossessor) {
        if (typeof newOwnerAndPossessor === "string") {
            this.newOwnerAndPossessor = new PublicKey_1.PublicKey(newOwnerAndPossessor);
        }
        else {
            this.newOwnerAndPossessor = newOwnerAndPossessor;
        }
        return this;
    }
    setAssetName(assetName) {
        if (typeof assetName === "string") {
            const utf8Encode = new TextEncoder();
            const nameBytes = utf8Encode.encode(assetName);
            this.assetName = new Uint8Array(8);
            nameBytes.forEach((b, i) => {
                this.assetName[i] = b;
            });
        }
        else {
            this.assetName = assetName;
        }
        return this;
    }
    getAssetName() {
        return this.assetName;
    }
    setNumberOfUnits(numberOfUnits) {
        if (typeof numberOfUnits === "number") {
            this.numberOfUnits = new Long_1.Long(numberOfUnits);
        }
        else {
            this.numberOfUnits = numberOfUnits;
        }
        return this;
    }
    getPackageSize() {
        return this._internalPackageSize;
    }
    getPackageData() {
        const builder = new QubicPackageBuilder_1.QubicPackageBuilder(this.getPackageSize());
        builder.add(this.issuer);
        builder.add(this.newOwnerAndPossessor);
        builder.addRaw(this.assetName);
        builder.add(this.numberOfUnits);
        return builder.getData();
    }
    getTransactionPayload() {
        const payload = new DynamicPayload_1.DynamicPayload(this.getPackageSize());
        payload.setPayload(this.getPackageData());
        return payload;
    }
}
exports.QubicTransferAssetPayload = QubicTransferAssetPayload;
